File: web/domain/filing/models/income/house-property/property-model.ts
import { HousePropertyLoanModel } from "./house-property-loan-model"
import { HousePropertyTenantModel } from "./house-property-tenent-model"
import { HousePropertyCoownerModel } from "./house-property-coowner-model"
import { HousePropertyAddressModel } from "./house-property-address-model"
import { HousePropertyModel } from "./house-property-model"

export interface PropertyModel {
    property: HousePropertyModel
    propertyAddress?: HousePropertyAddressModel | null
    propertyLoan?: HousePropertyLoanModel | null
    propertyTenants: HousePropertyTenantModel[]
    propertyCoowners: HousePropertyCoownerModel[]
  }
  export interface HousePropertyMasterData {
    states?: Array<{ value: string; label: string }>
    countries?: Array<{ value: string; label: string }>
    propertyTypes?: Array<{ value: string; label: string }>
    tenantIdentifierTypes?: Array<{ value: string; label: string }>
    coownerRelationships?: Array<{ value: string; label: string }>
    lenderTypes?: Array<{ value: string; label: string }> | null
  }
  
  /**
   * API Data Interfaces (camelCase)
   */
  export interface HousePropertyData {
    property: {
      propertyId?: number
      filingId?: number
      propertyType?: string
      ownershipShare?: number
      annualRentReceived?: number
      municipalTaxesPaid?: number
    }
    propertyAddress: {
      propertyAddressId?: number
      filingId?: number
      propertyId?: number
      addressLine1?: string
      addressLine2?: string
      city?: string
      district?: string
      state?: string
      postalCode?: string
      country?: string
    }
    propertyLoan?: {
      loanId?: number
      filingId?: number
      propertyId?: number
      vendorType?: string
      lenderName?: string
      loanAccountNumber?: string
      loanSanctionDate?: string  // YYYY-MM-DD
      totalLoanAmount?: number
      loanOutstanding?: number
      interestPaid?: number
      principalRepaid?: number
    }
    propertyTenants?: Array<{
      tenantId?: number
      filingId?: number
      propertyId?: number
      tenantName?: string
      identifierType?: string
      identifierValue?: string
    }>
    propertyCoowners?: Array<{
      coownerId?: number
      filingId?: number
      propertyId?: number
      coownerName?: string
      coownerPan?: string
      coownerRelationship?: string
      ownershipShare?: number
    }>
  }
  
  
  /**
   * Component Props
   */
  export interface HousePropertyWidgetProps {
    isOpen: boolean
    onClose: () => void
    filingId: number | null
    entityId?: number | string
    initialData?: Record<string, any> | null
    onSuccess?: (action: 'created' | 'updated', data: any) => void
    errorField?: string | null
    errorMessage?: string | null  // Error message to display for the field
    fieldErrors?: Record<string, string> | null  // Multiple field errors: { fieldName: errorMessage }
  }
  
  /**
   * Initial Form Data Constants
   */

File: api/filing/models/property_model.py
from .api_base_model import ApiBaseModel
from .house_property_model import HousePropertyModel
from .house_property_address_model import HousePropertyAddressModel
from .house_property_loan_model import HousePropertyLoanModel
from .house_property_tenant_model import HousePropertyTenantModel
from .house_property_coowner_model import HousePropertyCoownerModel


class PropertyModel(ApiBaseModel):
    property: HousePropertyModel
    property_address: HousePropertyAddressModel | None = None
    property_loan: HousePropertyLoanModel | None = None
    property_tenants: list[HousePropertyTenantModel] = []
    property_coowners: list[HousePropertyCoownerModel] = []

    @property
    def itr_loan_tkn_from(self) -> str:
        loan = self.property_loan
        vendor_type = (loan.vendor_type or "").lower() if loan else ""
        return "B" if "bank" in vendor_type else "I"

    @property
    def itr_bank_or_instn_name(self) -> str:
        loan = self.property_loan
        return ((loan.lender_name if loan else None) or "Bank")[:125] or "Bank"

    @property
    def itr_loan_acc_ref(self) -> dict:
        return {}

    @property
    def itr_dateofloan(self) -> str:
        loan = self.property_loan
        d = loan.loan_sanction_date if loan else None
        return d.strftime("%Y-%m-%d") if d else "2024-04-01"

    @property
    def itr_total_loan_amt(self) -> int:
        loan = self.property_loan
        return int(loan.total_loan_amount or 0) if loan else 0

    @property
    def itr_loan_outstanding_amt(self) -> int:
        loan = self.property_loan
        return int(loan.loan_outstanding or 0) if loan else 0

    @property
    def itr_interest_us24b(self) -> int:
        loan = self.property_loan
        return int(loan.interest_paid or 0) if loan else 0
# PropertyModel
